Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs

```diff
@@ -125,82 +124,54 @@ public bool DisconnectDevice(EntityUid uid, DeviceNetworkComponent? device, bool
-            if (_networks.TryGetValue(netId, out var network) && network.ReceiveAllDevices.TryGetValue(frequency, out var devices))
-                return devices;
-
-            return new();
+            return GetNetwork(device.DeviceNetId).Remove(device);
         }
 
         /// <summary>
         ///     Try to find a device on a network using its address.
         /// </summary>
-        private bool TryGetDevice(ConnectionType netId, string address, [NotNullWhen(true)] out DeviceNetworkComponent? device)
-        {
-            if (!_networks.TryGetValue(netId, out var network))
-            {
-                device = null;
-                return false;
-            }
-
-            return network.Devices.TryGetValue(address, out device);
-        }
-        #endregion
+        private bool TryGetDevice(ConnectionType netId, string address, [NotNullWhen(true)] out DeviceNetworkComponent? device) =>
+            GetNetwork(netId).Devices.TryGetValue(address, out device);
 
-        #region Packet Sending
         private void SendPacket(DeviceNetworkPacketEvent packet)
         {
-            HashSet<DeviceNetworkComponent> recipients;
-
+            var network = GetNetwork(packet.NetId);
             if (packet.Address == null)
             {
-                // Broadcast to all listening devices
-                recipients = GetListeningDevices(packet.NetId, packet.Frequency);
+                if (network.ListeningDevices.TryGetValue(packet.Frequency, out var devices))
+                {
+                    var deviceCopy = ArrayPool<DeviceNetworkComponent>.Shared.Rent(devices.Count);
+                    devices.CopyTo(deviceCopy);
+                    SendToConnections(deviceCopy, packet);
+                    ArrayPool<DeviceNetworkComponent>.Shared.Return(deviceCopy);
+                }
             }
             else
             {
-                // Add devices listening to all messages
-                recipients = new(GetRecieveAllDevices(packet.NetId, packet.Frequency));
-
-                // add the intended recipient (if they are even listening).
+                var totalDevices = 0;
+                if (network.ReceiveAllDevices.TryGetValue(packet.Frequency, out var devices))
+                {
+                    totalDevices += devices.Count;
+                }
                 if (TryGetDevice(packet.NetId, packet.Address, out var device) && device.ReceiveFrequency == packet.Frequency)
-                    recipients.Add(device);
+                {
+                    totalDevices += 1;
+                }
+                var deviceCopy = ArrayPool<DeviceNetworkComponent>.Shared.Rent(totalDevices);
+                if(devices != null)
+                {
+                    devices.CopyTo(deviceCopy);
+                }
+                if(device != null)
+                {
+                    deviceCopy[totalDevices - 1] = device;
+                }
+                SendToConnections(deviceCopy, packet);
```

Comment:
So I'm not very familiar with array pools, but my understanding is that deviceCopy will be some array with a length of at least totalDevices, but may be longer & have devices from some older rented array. So AFAIK you need to pass totalDevices to SendToConnections and ensure that it only iterates over the elements of deviceCopy that it should actually be sending things to?

Reply:
Not sure if it can have old ones either, but the size is now handled